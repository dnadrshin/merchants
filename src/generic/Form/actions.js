import keyMirror from 'keymirror';

const
  types = keyMirror({
    CREATE_FORM : null,
    CHANGE_FIELD: null,
    RESET_FORM  : null,
  }),

  createForm = (model, data) => ({model, data, type: types.CREATE_FORM}),
  changeField = (model, value) => ({model, value, type: types.CHANGE_FIELD}),
  resetForm = model => ({model, type: types.RESET_FORM});

export default {
  createForm, changeField, resetForm, types,
};
