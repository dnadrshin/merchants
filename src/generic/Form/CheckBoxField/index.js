// @flow
import _ from 'lodash';
import React from 'react';
import {compose} from 'recompose';
import {connect} from 'react-redux';
import actions from '../actions';

const
  CheckBoxField = (props: {
    changeField: ()=>{},
    model: string,
    lable: string,
    name: string,
    value: boolean,
  }) => <div className="checkbox">
    <label>
        <input
          type="checkbox"
          checked={props.checked}
          onChange={e => props.changeField(props.model, Boolean(e.target.checked))}
        /> {props.lable}
    </label>
  </div>;

export default compose(
  connect(
    (state, props) => ({
      checked: _.get(state.form, props.model, false)
    }),

    ({
      changeField: actions.changeField,
    })
  ),
)(CheckBoxField);
