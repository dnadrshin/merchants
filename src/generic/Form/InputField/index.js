// @flow
import _ from 'lodash';
import React from 'react';
import {compose} from 'recompose';
import {connect} from 'react-redux';
import actions from '../actions';

const
  InputField = (props: {
    changeField: (string, boolean)=>{},
    model: string,
    lable: string,
    name: string,
    value: string,
    placeholder: string,
    type: string
  }) => <div className="form-group">
    <label htmlFor={`label-${props.model}-${props.name}`}>{props.lable}</label>

    <input
      id={`label-${props.model}-${props.name}`}
      type={props.type}
      className="form-control"
      placeholder={props.placeholder}
      name={props.name}
      value={props.value}
      onChange={e => props.changeField(props.model, e.target.value)}
    />
  </div>;

export default compose(connect(
  (state, props) => ({
    value: _.get(state.form, props.model, ''),
  }),

  ({
    changeField: actions.changeField,
  }),
))(InputField);
