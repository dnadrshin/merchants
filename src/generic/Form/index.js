// @flow
import React from 'react';
import {connect} from 'react-redux';
import {lifecycle, compose} from 'recompose';
import actions from './actions';

const
  Form = (props: {
    children?: React$Element<*>
  }) => <form>
    {props.children}
  </form>;

export default compose(
  connect(
    null,

    ({
      createForm: actions.createForm,
      resetForm : actions.resetForm,
    }),
  ),

  lifecycle({
    componentDidMount() {
      this.props.createForm(this.props.model, this.props.data);
    },

    componentWillUnmount() {
      this.props.resetForm(this.props.model);
    },
  }),
)(Form);
