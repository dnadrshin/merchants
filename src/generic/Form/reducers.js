import _ from 'lodash';
import actions from './actions';
import {reduce} from '../helpers';

export default (state = {}, action) => reduce(state, action, {

  [actions.types.CREATE_FORM]: () => ({
    ...state,

    [action.model]: {
      ...state[action.model],
      ...action.data,
    },
  }),

  [actions.types.CHANGE_FIELD]: () => ({
    ..._.set(Object.assign(state), action.model, action.value)
  }),


  [actions.types.RESET_FORM]: () => ({    ...state,

    [action.model]: {},
  }),
});
