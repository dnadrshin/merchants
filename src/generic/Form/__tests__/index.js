import React from 'react';
import Form from '../';
import InputField from '../InputField';
import CheckBoxField from '../CheckBoxField';
import {createStore, combineReducers} from 'redux';
import formReducer from '../reducers';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme from 'enzyme';
import { Provider } from 'react-redux';

Enzyme.configure({ adapter: new Adapter() })

const
  data = {
    id: '1',
    firstname: 'sdfdsf',
    lastname: 'sdfsdf',
    avatarUrl: 'http://sdsfd.com',
    email: 'sdfsf@sdsdf.com',
    phone: '34534345',
    hasPremium: true,
    bids: [
      {
        id: '1',
        carTitle: 'BMW',
        amount: 10,
        created: 'somedate'
      },
      {
        id: '2',
        carTitle: 'VW',
        amount: 10,
        created: 'somedate'
      },
      {
        id: '3',
        carTitle: 'Mers',
        amount: 10,
        created: 'somedate'
      },
    ]
  },

  store = createStore(combineReducers({
    form: formReducer,
  })),

  instance = Enzyme.mount(<Provider store={store}>
    <Form model="editModal" data={data} >
      <InputField model="editModal.phone" name="phone" lable="Phone" />
      <CheckBoxField model="editModal.hasPremium" lable="Premium" />
    </Form>
  </Provider>);


describe('Form component', () => {
  it('render form with two Field with proper values', () => {
    expect(toJson(instance)).toMatchSnapshot();
    expect(instance.find('input[name="phone"]').props().value).toBe('34534345');
    expect(instance.find('input[type="checkbox"]').props().checked).toBe(true);
  });

  it('changed field value changing store', () => {
    instance.find('input[name="phone"]').simulate('change', {target: {value: '+1111122222'}});
    instance.find('input[type="checkbox"]').simulate('change', {target: {checked: false}});
    expect(toJson(instance)).toMatchSnapshot();
    expect(store.getState().form.editModal.phone).toBe('+1111122222');
    expect(store.getState().form.editModal.hasPremium).toBe(false);
  });
});
