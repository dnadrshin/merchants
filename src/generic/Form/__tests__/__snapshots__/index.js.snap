// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component changed field value changing store 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(lifecycle(Form))
    data={
      Object {
        "avatarUrl": "http://sdsfd.com",
        "bids": Array [
          Object {
            "amount": 10,
            "carTitle": "BMW",
            "created": "somedate",
            "id": "1",
          },
          Object {
            "amount": 10,
            "carTitle": "VW",
            "created": "somedate",
            "id": "2",
          },
          Object {
            "amount": 10,
            "carTitle": "Mers",
            "created": "somedate",
            "id": "3",
          },
        ],
        "email": "sdfsf@sdsdf.com",
        "firstname": "sdfdsf",
        "hasPremium": true,
        "id": "1",
        "lastname": "sdfsdf",
        "phone": "34534345",
      }
    }
    model="editModal"
  >
    <lifecycle(Form)
      createForm={[Function]}
      data={
        Object {
          "avatarUrl": "http://sdsfd.com",
          "bids": Array [
            Object {
              "amount": 10,
              "carTitle": "BMW",
              "created": "somedate",
              "id": "1",
            },
            Object {
              "amount": 10,
              "carTitle": "VW",
              "created": "somedate",
              "id": "2",
            },
            Object {
              "amount": 10,
              "carTitle": "Mers",
              "created": "somedate",
              "id": "3",
            },
          ],
          "email": "sdfsf@sdsdf.com",
          "firstname": "sdfdsf",
          "hasPremium": true,
          "id": "1",
          "lastname": "sdfsdf",
          "phone": "34534345",
        }
      }
      model="editModal"
      resetForm={[Function]}
    >
      <Form
        createForm={[Function]}
        data={
          Object {
            "avatarUrl": "http://sdsfd.com",
            "bids": Array [
              Object {
                "amount": 10,
                "carTitle": "BMW",
                "created": "somedate",
                "id": "1",
              },
              Object {
                "amount": 10,
                "carTitle": "VW",
                "created": "somedate",
                "id": "2",
              },
              Object {
                "amount": 10,
                "carTitle": "Mers",
                "created": "somedate",
                "id": "3",
              },
            ],
            "email": "sdfsf@sdsdf.com",
            "firstname": "sdfdsf",
            "hasPremium": true,
            "id": "1",
            "lastname": "sdfsdf",
            "phone": "34534345",
          }
        }
        model="editModal"
        resetForm={[Function]}
      >
        <form>
          <Connect(InputField)
            lable="Phone"
            model="editModal.phone"
            name="phone"
          >
            <InputField
              changeField={[Function]}
              lable="Phone"
              model="editModal.phone"
              name="phone"
              value="+1111122222"
            >
              <div
                className="form-group"
              >
                <label
                  htmlFor="label-editModal.phone-phone"
                >
                  Phone
                </label>
                <input
                  className="form-control"
                  id="label-editModal.phone-phone"
                  name="phone"
                  onChange={[Function]}
                  value="+1111122222"
                />
              </div>
            </InputField>
          </Connect(InputField)>
          <Connect(CheckBoxField)
            lable="Premium"
            model="editModal.hasPremium"
          >
            <CheckBoxField
              changeField={[Function]}
              checked={false}
              lable="Premium"
              model="editModal.hasPremium"
            >
              <div
                className="checkbox"
              >
                <label>
                  <input
                    checked={false}
                    onChange={[Function]}
                    type="checkbox"
                  />
                   
                  Premium
                </label>
              </div>
            </CheckBoxField>
          </Connect(CheckBoxField)>
        </form>
      </Form>
    </lifecycle(Form)>
  </Connect(lifecycle(Form))>
</Provider>
`;

exports[`Form component render form with two Field with proper values 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(lifecycle(Form))
    data={
      Object {
        "avatarUrl": "http://sdsfd.com",
        "bids": Array [
          Object {
            "amount": 10,
            "carTitle": "BMW",
            "created": "somedate",
            "id": "1",
          },
          Object {
            "amount": 10,
            "carTitle": "VW",
            "created": "somedate",
            "id": "2",
          },
          Object {
            "amount": 10,
            "carTitle": "Mers",
            "created": "somedate",
            "id": "3",
          },
        ],
        "email": "sdfsf@sdsdf.com",
        "firstname": "sdfdsf",
        "hasPremium": true,
        "id": "1",
        "lastname": "sdfsdf",
        "phone": "34534345",
      }
    }
    model="editModal"
  >
    <lifecycle(Form)
      createForm={[Function]}
      data={
        Object {
          "avatarUrl": "http://sdsfd.com",
          "bids": Array [
            Object {
              "amount": 10,
              "carTitle": "BMW",
              "created": "somedate",
              "id": "1",
            },
            Object {
              "amount": 10,
              "carTitle": "VW",
              "created": "somedate",
              "id": "2",
            },
            Object {
              "amount": 10,
              "carTitle": "Mers",
              "created": "somedate",
              "id": "3",
            },
          ],
          "email": "sdfsf@sdsdf.com",
          "firstname": "sdfdsf",
          "hasPremium": true,
          "id": "1",
          "lastname": "sdfsdf",
          "phone": "34534345",
        }
      }
      model="editModal"
      resetForm={[Function]}
    >
      <Form
        createForm={[Function]}
        data={
          Object {
            "avatarUrl": "http://sdsfd.com",
            "bids": Array [
              Object {
                "amount": 10,
                "carTitle": "BMW",
                "created": "somedate",
                "id": "1",
              },
              Object {
                "amount": 10,
                "carTitle": "VW",
                "created": "somedate",
                "id": "2",
              },
              Object {
                "amount": 10,
                "carTitle": "Mers",
                "created": "somedate",
                "id": "3",
              },
            ],
            "email": "sdfsf@sdsdf.com",
            "firstname": "sdfdsf",
            "hasPremium": true,
            "id": "1",
            "lastname": "sdfsdf",
            "phone": "34534345",
          }
        }
        model="editModal"
        resetForm={[Function]}
      >
        <form>
          <Connect(InputField)
            lable="Phone"
            model="editModal.phone"
            name="phone"
          >
            <InputField
              changeField={[Function]}
              lable="Phone"
              model="editModal.phone"
              name="phone"
              value="34534345"
            >
              <div
                className="form-group"
              >
                <label
                  htmlFor="label-editModal.phone-phone"
                >
                  Phone
                </label>
                <input
                  className="form-control"
                  id="label-editModal.phone-phone"
                  name="phone"
                  onChange={[Function]}
                  value="34534345"
                />
              </div>
            </InputField>
          </Connect(InputField)>
          <Connect(CheckBoxField)
            lable="Premium"
            model="editModal.hasPremium"
          >
            <CheckBoxField
              changeField={[Function]}
              checked={true}
              lable="Premium"
              model="editModal.hasPremium"
            >
              <div
                className="checkbox"
              >
                <label>
                  <input
                    checked={true}
                    onChange={[Function]}
                    type="checkbox"
                  />
                   
                  Premium
                </label>
              </div>
            </CheckBoxField>
          </Connect(CheckBoxField)>
        </form>
      </Form>
    </lifecycle(Form)>
  </Connect(lifecycle(Form))>
</Provider>
`;
